### DISCUSSION TITLE
Improving initialization speed

### DISCUSSION THREAD

**Discussion by PhotoKevin** [RomanIakovlev/timeshape#137] [Q&A]
Is there any way to improve the initialization speed? I'm trying to use TimeShape in an Android app but it takes 12-15 seconds to initialize the entire planet. I have that in the background but my app isn't really useful until it finishes and that's an eternity. 

I tried limiting it to a quarter of the globe but that made zero impact on the initialization time.

**Comment by RomanIakovlev** [comment]
That's right, currently reducing the bounding box doesn't have impact on initialization speed, only on the total memory usage. I'd be open to suggestions on how this can be improved though. 

E.g. one thing we could do is to split the Timeshape data into multiple geo-based parts (tiles), and only load those that are covered by the bounding box. Right now it's all stored in a single large file, which is read in its entirety. However this is a non-trivial amount of work, so if you're interested in making it happen, I'd gladly accept a pull request.

**Reply by PhotoKevin** [reply]
I need to work up a fork and create a PR, but I found that implementing serializable gets me where I want to be. Allow selecting a region, the first initialization is slow, sure, but I can serialize that into my cache directory. Deserializing the cache typically takes less than one second for North America.

The change is to add these two methods to TimeZoneEngine and add a couple "implements serializable" 

```
   public void serialize (File f) throws IOException
   {
      FileOutputStream fileOutputStream = new FileOutputStream (f, false);
      try (ObjectOutputStream objectOutputStream = new ObjectOutputStream (new BufferedOutputStream (fileOutputStream)))
      {
         objectOutputStream.writeObject (this.index);
         objectOutputStream.flush ();
      }
   }
   @SuppressWarnings("unchecked")
   public static TimeZoneEngine deSerialize (File f) throws IOException, ClassNotFoundException
   {
      FileInputStream fileInputStream = new FileInputStream(f);
      Index i;
      try (ObjectInputStream objectInputStream = new ObjectInputStream(fileInputStream))
      {
         return new TimeZoneEngine ((Index) objectInputStream.readObject ());
      }
   }
```

**Reply by PhotoKevin** [reply]
Sigh, how did that Index i come back from the dead.

**Reply by RomanIakovlev** [reply]
Interesting! Just so that I understand, do you serialize the initialized classes, and then attempt to read them back on the next start of your app / activity, falling back to the standard initialization if the serialized data is missing?

**Reply by PhotoKevin** [reply]
Exactly. It's kind of a mess right now in my code, as it runs through background threads, all of the try/catches, and logging, so I'll just give psuedo code:

This is called from my onCreate method:
```
File f = ... // Ends up being something like TZ_NorthAmerica.cache in the app cache directory.
if (f.exists())
   engine = TimeZoneEngine.deserialize (f); // Real code has this in a try/catch
if (engine == null) 
{
   engine = TimeZoneEngine.initialize (...) // Bounding box supplied from user preferences. 
   engine.serialize (f); // Saved for next time the app runs.
}
```
The first time the app is started, or when the user selects a different region from settings, it won't find the cached data. In that case I look up the bounding box for the region from a JSON file and initialize using that area. It also does the slow initialization if deserializing fails. Maybe because the serialized data is from an earlier version of TimeShape.

**Reply by PhotoKevin** [reply]
If you like this, I'll get a fork and PR together.

**Reply by RomanIakovlev** [reply]
I like this, it would be great if you could contribute it! I'd please ask you to create an issue first, to describe what exactly you're planning to add or change, as well as some code examples of how this new feature can be used.

**Reply by PhotoKevin** [reply]
I've created an issue.
